expo/expo
# [autolinking][build-properties] support 3rd party reuse for extraPods

# Why

people would like to reuse with `withBuildProperties()` for adding extraPods or extraMavenRepos. originally the extraPods or  extraMavenRepos are only supported from app config. library authors with their own config-plugins are not supported.
follows up https://github.com/expo/expo/issues/24099#issuecomment-1697585578

also fixes #25455 and close ENG-10696

# How

- [expo-build-properties] serialize the extraPods/extraMavenRepos to Podfile.properties.json/gradle.properties
- [expo-modules-autolinking] read the extra dependencies from Podfile.properties.json/gradle.properties

# Test Plan

1. add the local config-plugin for bare-expo
```js
const { withBuildProperties } = require('expo-build-properties');

const withExtraBuildDeps = (config) => {
  return withBuildProperties(config, {
    android: { extraMavenRepos: ['https://customers.pspdfkit.com/maven/'] },
    ios: { extraPods: [{ name: 'Protobuf' }] },
  });
};

module.exports = withExtraBuildDeps;
```

2. `npx expo prebuild --clean`
3.  check the contents in Podfile.properties.json and gradle.properties
4. on ios, pod install and check "Protobuf" is added to Podfile.lock
5. on android, adding `implementation 'com.pspdfkit:libraries-java:1.4.1'` to **android/app/build.gradle** and try to build bare-expo

# Checklist
